CCBoot.js

```diff
@@ -184,168 +184,141 @@ cc.isCrossOrigin = function (url) {
+     * Each function will be passed to the next function
+     * If an error occurred, next function would not be executed
+     * @param {Array} tasks
      * @param {function} cb callback
      * @param {Object} [target]
-     * @return {cc.AsyncPool}
      */
     waterfall : function(tasks, cb, target){
-        var args = [];
-        var lastResults = [null];//the array to store the last results
-        var asyncPool = new cc.AsyncPool(tasks, 1,
-            function (func, index, cb1) {
-                args.push(function (err) {
-                    args = Array.prototype.slice.call(arguments, 1);
-                    if(tasks.length - 1 === index) lastResults = lastResults.concat(args);//while the last task
-                    cb1.apply(null, arguments);
-                });
-                func.apply(target, args);
-            }, function (err) {
-                if (!cb)
-                    return;
-                if (err)
-                    return cb.call(target, err);
-                cb.apply(target, lastResults);
-            });
-        asyncPool.flow();
-        return asyncPool;
+        var index = 0;
+        var func;
+        var callback = function(error){
+            var args;
+            args = Array.prototype.slice.call(arguments, 1);
+            args.push(callback);
+            func = tasks[++index];
+            if(!error && index < tasks.length && func){
+                func.apply(this, args);
+            }else{
+                cb.apply(target || this, arguments);
+            }
+        };
+        func = tasks[0];
+        if(func)
+            func(callback);
+        else
+            callback(null);
     },
 
     /**
-     * Do tasks by iterator.
+     * Similar to parallel
+     * Each task is processed by the incoming iterator
      * @param {Array|Object} tasks
      * @param {function|Object} iterator
-     * @param {function} [callback]
+     * @param {function} [cb]
      * @param {Object} [target]
-     * @return {cc.AsyncPool}
-     */
-    map : function(tasks, iterator, callback, target){
-        var locIterator = iterator;
-        if(typeof(iterator) === "object"){
-            callback = iterator.cb;
-            target = iterator.iteratorTarget;
-            locIterator = iterator.iterator;
+     */
+    map : function(tasks, iterator, cb, target){
+        var index = 0, isArray = Array.isArray(tasks);
+        var values = isArray ? [] : {};
+
+        var handle = function(p){
+            if(isArray) p = p - 0;
+            var task = tasks[p];
+            if(task){
+                values[p] = undefined;
+                setTimeout(function(){
+                    iterator(task, p, function(error, value){
+                        values[p] = value;
+                        index--;
+                        if(error || index <= 0){
+                            index = Number.POSITIVE_INFINITY;
+                            cb.call(target || this, error, values);
```

Comment:
This may be invoked multiple times, because calling the callback doesn't stop other handles function's execution. So other resources will continue to load, and once another error happens, it will call the callback once again. Or when the index reduce to 0, it will also call the callback